cmd/prometheus/main.go

```diff
@@ -244,8 +245,22 @@ func main() {
 	// RoutePrefix must always be at least '/'.
 	cfg.web.RoutePrefix = "/" + strings.Trim(cfg.web.RoutePrefix, "/")
 
+	// Check for overflows. This limits our max retention to ~292.5y.
+	if cfg.tsdb.Retention < 0 {
+		cfg.tsdb.Retention = math.MaxInt64
+	}
+
 	if cfg.tsdb.MaxBlockDuration == 0 {
 		cfg.tsdb.MaxBlockDuration = cfg.tsdb.Retention / 10
+
+		weekLong, err := model.ParseDuration("7d")
+		if err != nil {
+			panic(err)
+		}
+
+		if cfg.tsdb.MaxBlockDuration > weekLong {
+			cfg.tsdb.MaxBlockDuration = weekLong
```

Comment:
How was this value determined? I'd be looking at something in the one month to one year range.

Reply:
That was before the memory improvements I made in the last release. Having too small blocks will actually increase memory usage.

Reply:
I still see 20% bump in memory compacting 7d blocks on a typical server with 100k samples/sec.

What is the the overhead of having more blocks?

Reply:
It's duplicating the index memory structures, plus the various other bits and pieces. It'll also slow down and require more memory for queries across blocks.

Reply:
Right, but how much is that in terms of bytes or percent of server? I see a nearly linear percent burst based on the size of the compaction, so if we went with 28d blocks on my server, I would see an 80% burst in memory use on compaction.

Reply:
With the current code, I'd expect it to always be around 10% as it'd be creating an index for up to 10% of data.